File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    boolean [] visited;
    public void dfs(int [][] isConnected  , int R){
        for(int i=0;i<isConnected.length ; i++){
            if(isConnected[R][i] ==1 && !visited[i]){
                visited[i]= true;
                dfs(isConnected , i);
            }
        }
    }
    public int findCircleNum(int[][] isConnected) {
        int N = isConnected.length;
        visited = new boolean[N];
        int count = 0;
        for(int row=0;row<N;row++){
            for(int col=0;col<N;col++){
                if(isConnected[row][col]==1 && !visited[row] ){
                count++;
                visited[row] = true;
                dfs(isConnected ,row);
            }
            }

        }
        return count;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    vector<int> leader;
    int find(int node){
        if(node !=leader[node])
            leader[node] = find(leader[node]);
        return leader[node];
    }
    void join(int left , int right){
        leader[find(right)] = find(left);
    }
    
    int findCircleNum(vector<vector<int>>& isConnected) {
        int N = isConnected.size();
         leader.resize(N+1 );
        for(int i=1;i<=N;i++){
            leader[i] = i;
        }
        for(int row =0;row<N;row++){
            for(int col =0;col<N;col++){
                if(isConnected[row][col]==1){
                    join(row+1 ,col+1);
                }
            }
        }
        set<int> s;
        for(int i=1;i<=N;i++){
            s.insert(find(leader[i]));
        }
        return s.size();
    }
};